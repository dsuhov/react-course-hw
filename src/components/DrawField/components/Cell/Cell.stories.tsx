import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, number, boolean } from "@storybook/addon-knobs";
import { Cell } from "./Cell";
export default {
  title: "Cell",
  decorators: [withKnobs],
};

export const nonAlive = () => (
  <Cell
    isAlive={boolean("Alive or Dead", false)}
    clickHandler={action("Cell clicked")}
    x={number("x", 10)}
    y={number("y", 23)}
  />
);

export const nonAliveWithZeroX = () => (
  <Cell
    isAlive={boolean("Alive or Dead", false)}
    clickHandler={action("Cell clicked")}
    x={number("x", 0)}
    y={number("y", 23)}
  />
);

export const nonAliveWithZeroY = () => (
  <Cell
    isAlive={boolean("Alive or Dead", false)}
    clickHandler={action("Cell clicked")}
    x={number("x", 12)}
    y={number("y", 0)}
  />
);

export const nonAliveWithZeroXY = () => (
  <Cell
    isAlive={boolean("Alive or Dead", false)}
    clickHandler={action("Cell clicked")}
    x={number("x", 0)}
    y={number("y", 0)}
  />
);

export const Alive = () => (
  <Cell
    isAlive={boolean("Alive or Dead", true)}
    clickHandler={action("Cell clicked")}
    x={number("x", 10)}
    y={number("y", 23)}
  />
);
