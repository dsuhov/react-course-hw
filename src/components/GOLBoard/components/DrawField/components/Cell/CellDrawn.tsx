import styled from "@emotion/styled";
import { css, keyframes } from "@emotion/core";

interface CellDrawnProps {
  isAlive: boolean;
  x: number;
  y: number;
}

const borderColor = "rgb(85, 74, 58)";
const dyingBg = "background-color: rgb(250, 42, 12)";
const revivingBg = "background-color: rgb(59, 197, 0)";
const deadBg = "background-color: #fff";
const aliveBg = "background-color: #050505";

const animateTransitionDied = keyframes`
  0% {
    ${aliveBg};
  }
  50% {
    ${dyingBg};
  }
  100% {
    ${deadBg};
  }
`;

const animateTransitionRevived = keyframes`
  0% {
    ${deadBg};
  }
  50% {
    ${revivingBg};
  }
  100% {
    ${aliveBg};
  }
`;

const MainCell = css`
  width: 10px;
  height: 10px;
  border-right: 1px solid ${borderColor};
  border-bottom: 1px solid ${borderColor};
  flex: 0 0 10px;
`;

const DeadCell = css`
  ${deadBg};
  animation: ${animateTransitionDied} 0.1s linear;
`;

const AliveCell = css`
  ${aliveBg};
  animation: ${animateTransitionRevived} 0.1s linear;
`;

export const CellDrawn = styled.div`
  ${MainCell};
  ${({ isAlive }: CellDrawnProps) => (isAlive ? AliveCell : DeadCell)};
  ${({ x }: CellDrawnProps) =>
    x === 0 ? `border-left: 1px solid ${borderColor}` : "border-left: none"};
  ${({ y }: CellDrawnProps) =>
    y === 0 ? `border-top: 1px solid ${borderColor}` : "border-top: none"};
`;
