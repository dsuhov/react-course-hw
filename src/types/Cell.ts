export interface CellProps {
  // x-coordinate of the cell
  x: number;
  // y-coordinate of the cell
  y: number;
  // status of the cell, dead of alive
  isAlive: boolean;
  // click handler to inform about coordinates of the cell to further actions
  clickHandler: (x: number, y: number) => void;
}

export interface CellDrawnProps {
  isAlive: boolean;
  x: number;
  y: number;
}
