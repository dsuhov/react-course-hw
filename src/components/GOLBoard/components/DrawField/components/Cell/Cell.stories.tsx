import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, number, boolean } from "@storybook/addon-knobs";
import { RawCell as Cell } from "./Cell";
export default {
  title: "Cell",
  decorators: [withKnobs],
};

import { Provider } from "react-redux";
import { store } from "@/rdx/store";

export const nonAlive = () => (
  <Provider store={store}>
    <Cell
      isAlive={boolean("Alive or Dead", false)}
      cellClick={action("Cell clicked")}
      x={number("x", 10)}
      y={number("y", 23)}
    />
  </Provider>
);

export const nonAliveWithZeroX = () => (
  <Provider store={store}>
    <Cell
      isAlive={boolean("Alive or Dead", false)}
      cellClick={action("Cell clicked")}
      x={number("x", 0)}
      y={number("y", 23)}
    />
  </Provider>
);

export const nonAliveWithZeroY = () => (
  <Provider store={store}>
    <Cell
      isAlive={boolean("Alive or Dead", false)}
      cellClick={action("Cell clicked")}
      x={number("x", 12)}
      y={number("y", 0)}
    />
  </Provider>
);

export const nonAliveWithZeroXY = () => (
  <Provider store={store}>
    <Cell
      isAlive={boolean("Alive or Dead", false)}
      cellClick={action("Cell clicked")}
      x={number("x", 0)}
      y={number("y", 0)}
    />
  </Provider>
);

export const Alive = () => (
  <Provider store={store}>
    <Cell
      isAlive={boolean("Alive or Dead", true)}
      cellClick={action("Cell clicked")}
      x={number("x", 10)}
      y={number("y", 23)}
    />
  </Provider>
);
