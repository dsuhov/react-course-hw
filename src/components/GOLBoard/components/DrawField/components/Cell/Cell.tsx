import React from "react";
import { CellDrawn } from "./CellDrawn";
import { connect } from "react-redux";
import { gameFieldActions } from "@/rdx/gameField/gameFieldSlice";

export interface CellProps {
  // x-coordinate of the cell
  x: number;
  // y-coordinate of the cell
  y: number;
  // status of the cell, dead of alive
  isAlive: boolean;
  cellClick: (coords: CellCorrds) => void;
}

export class RawCell extends React.PureComponent<CellProps> {
  render() {
    const { x, y } = this.props;

    return (
      <CellDrawn
        {...this.props}
        onClick={() => this.props.cellClick({ x, y })}
      />
    );
  }
}

const mapDispatchToProps = {
  cellClick: gameFieldActions.cellClick,
};

export const Cell = connect(null, mapDispatchToProps)(RawCell);
