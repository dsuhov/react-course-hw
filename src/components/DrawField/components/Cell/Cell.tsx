import React, { Component } from "react";
import { CellProps } from "types/Cell";
import { CellDrawn } from "./CellDrawn";

export class Cell extends Component<CellProps> {
  constructor(props: CellProps) {
    super(props);
  }

  shouldComponentUpdate(nextProps: CellProps) {
    return this.props.isAlive !== nextProps.isAlive;
  }

  render() {
    return (
      <CellDrawn
        {...this.props}
        onClick={() => this.props.clickHandler(this.props.x, this.props.y)}
      />
    );
  }
}