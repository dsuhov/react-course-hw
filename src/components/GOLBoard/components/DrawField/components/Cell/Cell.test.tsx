import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import { Cell } from "./Cell";

const deadCellProps = {
  x: 5,
  y: 5,
  isAlive: false,
};

const aliveCellProps = {
  x: 5,
  y: 5,
  isAlive: true,
};

describe("Cell", () => {
  it("renders dead cell", () => {
    expect(
      renderer
        .create(<Cell {...deadCellProps} clickHandler={jest.fn()} />)
        .toJSON()
    ).toMatchSnapshot();
  });

  it("renders dead cell", () => {
    expect(
      renderer
        .create(<Cell {...aliveCellProps} clickHandler={jest.fn()} />)
        .toJSON()
    ).toMatchSnapshot();
  });

  it("calls сlick callback on click by dead cell", () => {
    const onClick = jest.fn();
    const wrapper = mount(<Cell {...deadCellProps} clickHandler={onClick} />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });

  it("calls сlick callback on click by alive cell", () => {
    const onClick = jest.fn();
    const wrapper = mount(<Cell {...aliveCellProps} clickHandler={onClick} />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });

  it("calls click callback with passed x, y params", () => {
    const onClick = jest.fn();
    const x = 12;
    const y = 14;
    const wrapper = mount(
      <Cell clickHandler={onClick} x={x} y={y} isAlive={true} />
    );
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalledWith(x, y);
  });
});
