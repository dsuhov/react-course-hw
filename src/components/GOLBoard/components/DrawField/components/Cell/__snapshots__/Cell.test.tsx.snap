// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders alive cell 1`] = `
.emotion-0 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #050505;
  border-left: none;
  border-top: none;
}

<div
  className="emotion-0"
  onClick={[Function]}
  x={5}
  y={5}
/>
`;

exports[`Cell renders dead cell 1`] = `
.emotion-0 {
  width: 10px;
  height: 10px;
  border-right: 1px solid rgb(85,74,58);
  border-bottom: 1px solid rgb(85,74,58);
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: #fff;
  border-left: none;
  border-top: none;
}

<div
  className="emotion-0"
  onClick={[Function]}
  x={5}
  y={5}
/>
`;
